File: 0090-subsets-ii/solution.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        res = []
        nums.sort()
        def backtrack(index, path):
            if path not in res:
                res.append(path[:])

            for x in range(index, len(nums)):
                path.append(nums[x])
                backtrack(x + 1, path)
                path.pop()
    
        backtrack(0, [])  
        return res

            

File: 0090-subsets-ii/solution.cpp
class Solution {
public:
    void backtrack(vector<int>& path, vector<vector<int>>& res, int index, const vector<int>& nums){
        res.push_back(path);

        for(int i = index; i < nums.size(); ++i){
            if(index > nums.size()){
                return;
            }
            if((index < i) and (nums[i] == nums[i - 1])){
                continue;
            }
            path.push_back(nums[i]);
            backtrack(path, res, i + 1, nums);
            path.pop_back();
        }
    }
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> path;
        backtrack(path, res, 0, nums);
        return res;
    }
};
